File: cpp/Array/0034_Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
#include <vector>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  /*
   * binary search
   * [5, 7, 7, 8, 8, 10]
   *  |left          |right
   * find left bound and right bound separately
   * */
  vector<int> searchRange(vector<int>& nums, int target) {
    vector<int> res;
    int size = nums.size() - 1;
    int left, right;

    // find the left bound
    left = 0;
    right = size;
    while (left <= right) {
      int mid = left + ((right - left) >> 1);
      if (target > nums[mid]) left = mid + 1;
      else right = mid - 1; // shrink right bound
    }
    if (left <= size && nums[left] == target) res.emplace_back(left); // check left over size
    else res.emplace_back(-1);

    // find the right bound
    left = 0;
    right = size;
    while (left <= right) {
      int mid = left + ((right - left) >> 1);
      if (target < nums[mid]) right = mid - 1;
      else left = mid + 1; // shrink left bound
    }
    if (right >= 0 && nums[right] == target) res.emplace_back(right); // check right over size
    else res.emplace_back(-1);

    return res;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0034-Find-First-and-Last-Position-of-Element-in-Sorted-Array.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        leftbound = left

        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1
        rightbound = right
        return [leftbound, rightbound] if leftbound <= rightbound else [-1, -1]


s = Solution()
nums = [5, 7, 7, 8, 8, 10]
target = 8
print(s.searchRange(nums, target))

target = 6
print(s.searchRange(nums, target))

nums = []
target = 0
print(s.searchRange(nums, target))
